Simplify parsing utilities

eerovaher opened the pull request:
### Description

It is possible to simplify our parsing utilities by making use of `pathlib.Path` and the convenience methods it provides. Moving some common code that inserts a custom header to the files `ply` writes out from `lex()` and `yacc()` avoids code duplication, which simplifies the code even more. I'm not entirely convinced that moving that code to the same context manager that patches the `get_caller_module_dict()` functions is the best thing to do. An alternative would be to introduce a second context manager just for inserting the custom headers, but then we would end up with two context managers that are always called together, and that doesn't sound like a very good solution either.

The updated code is used by the unit parsers, so `units` maintainers might be interested in this.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk reviewed (approved):
This looks good! I assume you have tested that if you remove the various files with tables, they are indeed regenerated? (not sure we have a test for that...)

eerovaher commented:
I don't think we have an automated test that would check generating the files. The simplest way to check it manually is
```
$ rm astropy/units/format/generic_*
$ git status 
On branch simpler-parsing-utils
Your branch is up to date with 'eerovaher/simpler-parsing-utils'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    astropy/units/format/generic_lextab.py
	deleted:    astropy/units/format/generic_parsetab.py

no changes added to commit (use "git add" and/or "git commit -a")
$ pytest astropy/units/ > /dev/null
$ git status 
On branch simpler-parsing-utils
Your branch is up to date with 'eerovaher/simpler-parsing-utils'.

nothing to commit, working tree clean
```

mhvk commented:
Let's get this in - thanks @eerovaher!

eerovaher commented:
Turns out that the code that I modified is being executed in our tests after all: https://github.com/astropy/astropy/blob/00c8a65e31de7488184e77c557795d879ba58fe7/astropy/utils/tests/test_parsing.py. I didn't realize this because I tried to find relevant tests by modifying `astropy.utils.parsing._TAB_HEADER` and seeing which tests fail, but the tests in question explicitly rely on that private constant, so when I modified the code I also unknowingly modified the tests. Having tests that rely on private API is bad enough, but these tests rely on private implementation details (i.e. not just on the existence of the private constant, but also on its contents). In other words, these tests don't check that the code is doing what it is supposed to be doing, they instead check that the code is doing what the code is doing.

mhvk commented:
Not a great test indeed; I guess the purpose was mostly to see that one produced correct `lextab`, `parsetab` in the right locations.
